### PR TITLE
Fix a bug for imports run with different idScheme than UID

### PR DESCRIPTION
**Pull request by huzvak**
Fixes a bug with dataValues import when an idScheme different than UID was used.

DHIS2-7069

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/IdentifiableObjectUtils.java [hunk 1]
```diff
@@ -396,4 +395,27 @@ else if ( identifiableObject.getCode() != null && !identifiableObject.getCode().
 
         return object.getClass().getName();
     }
+
+    public static <T extends BaseIdentifiableObject> String getIdentifierBasedOnIdScheme( T object, IdScheme idScheme )
+    {
+        //idScheme with IdentifiableProperty.ATTRIBUTE has to be specially treated in the code dealing with Attribute IDs
```
Comment (stian-sandvold) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/IdentifiableObjectUtils.java]: Is it possible to move that logic into this method?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java [hunk 2]
```diff
@@ -1624,63 +1610,146 @@ private boolean doValidationOfMandatoryAttributes( User user )
...
-            // account also already stored data values in order to make correct decision. Basically, this situation happens when
-            // only 1 dataValue is updated and /events/{uid}/{dataElementUid} endpoint is leveraged.
-            if ( isSingleValueUpdate ) {
-                presentDataElements.addAll(
-                    programStageInstance.getDataValues().stream()
-                        .filter( dv -> !StringUtils.isEmpty( dv.getValue() ))
-                        .map( dv -> dv.getDataElement().getUid() )
-                        .collect( Collectors.toSet()));
-            }
+    private Map<String, String> validatePresenceOfMandatoryDataElementsWithOtherIdSchemes( Event event, ProgramStageInstance programStageInstance,
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java]: Is it possible to make this name a bit shorter? Like validateMandatoryDataElementForAttribute, etc?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java [hunk 3]
```diff
@@ -1624,63 +1610,146 @@ private boolean doValidationOfMandatoryAttributes( User user )
...
-                importSummary.setStatus( ImportStatus.ERROR );
-            }
+                DataElement dataElement = getDataElement( dataElementIdScheme, de );
+
+                if ( dataElement != null )
+                {
+                    presentDataElements.put( dataElement.getUid(), de );
+
+                    //If I have a correct Attribute:UID, then replace the Name byt this Attribute:UID. It will come handy later
+                    mandatoryDataElements.put( dataElement.getUid(), de );
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java]: Are you sure you should put it in mandatory and present and not just present? If yes, maybe mandatory should have a different name?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java [hunk 4]
```diff
@@ -1624,63 +1610,146 @@ private boolean doValidationOfMandatoryAttributes( User user )
...
+        // When the request is update, then only changed data values can be in the payload and so I should take into
+        // account also already stored data values in order to make correct decision. Basically, this situation happens when
+        // only 1 dataValue is updated and /events/{uid}/{dataElementUid} endpoint is leveraged.
+        if ( isSingleValueUpdate )
+        {
+            //Using Map<UID, Attribute:UID> to be able to make a Sets.difference below. Arbitrary Attribute:UID is used here as it is not significant
+            presentDataElements.putAll( programStageInstance.getDataValues().stream()
+                .filter( dv -> !StringUtils.isEmpty( dv.getValue() ) )
+                .map( dv -> dv.getDataElement().getUid() )
+                .collect( Collectors.toMap( Function.identity(), v -> "temp" ) ) );
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java]: "temp" can be just ""?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java [hunk 5]
```diff
@@ -1731,28 +1820,42 @@ private boolean validateDataValue( ProgramStageInstance programStageInstance, Us
...
+        if ( validateMandatoryAttributes )
         {
             ValidationStrategy validationStrategy = programStageInstance.getProgramStage().getValidationStrategy();
 
-            if ( (validationStrategy == ValidationStrategy.ON_UPDATE_AND_INSERT ||
-                (validationStrategy == ValidationStrategy.ON_COMPLETE && eventStatus == EventStatus.COMPLETED)) &&
-                (mandatoryDataElements.contains( dataElement.getUid() ) && (StringUtils.isEmpty( value ) || "null".equals( value ))) )
+            if ( ( validationStrategy == ValidationStrategy.ON_UPDATE_AND_INSERT ||
+                ( validationStrategy == ValidationStrategy.ON_COMPLETE && eventStatus == EventStatus.COMPLETED )) &&
+                ( mandatoryDataElements.containsKey( dataElement.getUid() ) && ( StringUtils.isEmpty( value ) || "null".equals( value ))) )
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java]: space between closing parantheses also :)

Comment (huzvak) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java]: Fixed

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java (hunk 4 above)
Comment (huzvak) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java]: Fixed

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java (hunk 3 above)
Comment (huzvak) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java]: It is correct as it is. Just obtaining Attribute:UID where I can so later (in `validateDataValue()`) I can eventually show a more informative error message to the user.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java (hunk 2 above)
Comment (huzvak) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java]: Yes. Improved a bit. Hope it is enough.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/IdentifiableObjectUtils.java (hunk 1 above)
Comment (huzvak) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/IdentifiableObjectUtils.java]: Not really. The logic will differ every time. The Attribute:UID has to be treated in a specific way. Right now I am not able to obtain Attribute:UID from DE UID. Also, the method is generic for all IdentifiableObjects. But if we figure out how to do a mapping for Attribute:UIDs, then sure, it should be moved here.

**Pull request merged**